modules/calc/src/main/java/com/opengamma/strata/calc/runner/CalculationFunctions.java

```diff
@@ -0,0 +1,221 @@
+/**
+ * Copyright (C) 2016 - present by OpenGamma Inc. and the OpenGamma group of companies
+ *
+ * Please see distribution for license.
+ */
+package com.opengamma.strata.calc.runner;
+
+import static com.opengamma.strata.collect.Guavate.toImmutableMap;
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import org.joda.beans.BeanDefinition;
+import org.joda.beans.ImmutableBean;
+import org.joda.beans.ImmutableValidator;
+import org.joda.beans.JodaBeanUtils;
+import org.joda.beans.MetaBean;
+import org.joda.beans.Property;
+import org.joda.beans.PropertyDefinition;
+import org.joda.beans.impl.light.LightMetaBean;
+
+import com.google.common.collect.ImmutableMap;
+import com.opengamma.strata.basics.CalculationTarget;
+import com.opengamma.strata.calc.runner.function.CalculationFunction;
+import com.opengamma.strata.collect.ArgChecker;
+
+/**
+ * The calculation functions.
+ * <p>
+ * This provides the complete set of functions that will be used in a calculation.
+ * Each {@link CalculationFunction} handles a specific type of {@link CalculationTarget},
+ * thus the functions are keyed in a {@code Map} by the target type {@code Class}.
+ */
+@BeanDefinition(style = "light")
+public final class CalculationFunctions implements ImmutableBean, Serializable {
+
+  /**
+   * An empty instance.
+   */
+  private static final CalculationFunctions EMPTY = new CalculationFunctions(ImmutableMap.of());
+
+  /**
+   * The functions, keyed by target type.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final ImmutableMap<Class<?>, CalculationFunction<?>> functions;
+
+  //-------------------------------------------------------------------------
+  /**
+   * Obtains an empty instance with no functions.
+   * 
+   * @return the empty instance
+   */
+  public static CalculationFunctions empty() {
+    return EMPTY;
+  }
+
+  /**
+   * Obtains an instance from the specified functions.
+   * <p>
+   * The list will be converted to a {@code Map} using {@link CalculationFunction#targetType()}.
+   * Each function must refer to a different target type.
+   * 
+   * @param functions  the functions
+   * @return the calculation functions
+   */
+  @SafeVarargs
+  public static CalculationFunctions of(CalculationFunction<?>... functions) {
+    return new CalculationFunctions(Stream.of(functions).collect(toImmutableMap(fn -> fn.targetType())));
```

Comment:
Just to note, you could use `MapStream` here:

    MapStream.of(functions, fn -> fn.targetType()).toMap();